presto-native-execution/presto_cpp/main/LinuxMemoryChecker.cpp

```diff
@@ -0,0 +1,184 @@
+  // consistency we will match cgroup V1 and change if
+  // necessary.
+  int64_t systemUsedMemoryBytes() override {
+    static std::string statFile;
+    static const boost::regex inactiveAnonRegex(
+        R"!(inactive_anon\s*(\d+)\s*)!");
+    static const boost::regex activeAnonRegex(R"!(active_anon\s*(\d+)\s*)!");
+    static const boost::regex memAvailableRegex(
+        R"!(MemAvailable:\s*(\d+)\s*kB)!");
+    static const boost::regex memTotalRegex(R"!(MemTotal:\s*(\d+)\s*kB)!");
+    static const char* cgroupV1Path = "/sys/fs/cgroup/memory/memory.stat";
+    static const char* cgroupV2Path = "/sys/fs/cgroup/memory.stat";
+
+    size_t memAvailable = 0;
+    size_t memTotal = 0;
+    size_t inactiveAnon = 0;
+    size_t activeAnon = 0;
+
+    // Find out what cgroup version (v1 or v2) we have based on the directory
+    // it's mounted.
+    if (statFile.empty()) {
+      struct stat buffer;
+      if ((stat(cgroupV1Path, &buffer) == 0)) {
+        statFile = cgroupV1Path;
+      } else if ((stat(cgroupV2Path, &buffer) == 0)) {
+        statFile = cgroupV2Path;
+      } else {
+        statFile = "None";
+      }
+      LOG(INFO) << fmt::format("Using memory stat file {}", statFile);
+    }
+
+    if (statFile != "None") {
+      folly::gen::byLine(statFile.c_str()) |
+          [&](const folly::StringPiece& line) -> void {
+        inactiveAnon =
+            extractNumericConfigValueWithRegex(line, inactiveAnonRegex);
+        activeAnon = extractNumericConfigValueWithRegex(line, activeAnonRegex);
+      };
+
+      // Unit is in bytes.
+      return inactiveAnon + activeAnon;
+    }
+
+    // Last resort use host machine info.
+    folly::gen::byLine("/proc/meminfo") |
+        [&](const folly::StringPiece& line) -> void {
+      memAvailable =
+          extractNumericConfigValueWithRegex(line, memAvailableRegex) * 1024;
+      memTotal = extractNumericConfigValueWithRegex(line, memTotalRegex) * 1024;
+    };
+    // Unit is in bytes.
+    return (memAvailable && memTotal) ? memTotal - memAvailable : 0;
+  }
+
+  int64_t mallocBytes() const override {
+    return mallocBytes_;
+  }
+
+  void periodicCb() const override {
+    if (periodicCb_) {
+      periodicCb_();
+    }
+  }
+
+  bool heapDumpCb(const std::string& filePath) const override {
+    if (heapDumpCb_) {
+      return heapDumpCb_(filePath);
+    }
+    return false;
+  }
+
+  void removeDumpFile(const std::string& filePath) const override {}
+
+ private:
+  int64_t mallocBytes_{0};
+  std::function<void()> periodicCb_;
+  std::function<bool(const std::string&)> heapDumpCb_;
+  size_t extractNumericConfigValueWithRegex(
+      const folly::StringPiece& line,
+      const boost::regex& regex) {
+    boost::cmatch match;
+    if (boost::regex_match(line.data(), match, regex)) {
+      if (match.size() > 1) {
+        std::string numStr(match[1].str());
+        return std::stoul(numStr);
+      }
+    }
+    return 0;
+  }
+};
+
+namespace {
```

Comment:
Can we add memoryLimitCheckerType and LinuxMemoryContainer is one type supported by OSS (We might have other types in the future). And initialize the memory limit checker in addMemoryCheckerPeriodicTask which could be virtual method in PrestoServer. Meta could overload this internally with our own backend implementation built for Meta infra? Thanks!

Reply:
@xiaoxmeng, @minhancao  We can do this similar to the stats reporter `PRESTO_STATS_REPORTER_TYPE`.
We are using a `folly::Singleton<facebook::presto::PeriodicMemoryChecker> checker` that Meta can set based on `PRESTO_MEMORY_CHECKER_TYPE`

Reply:
So it would look something like this? Was thinking of making the change below in LinuxMemoryChecker.cpp where my LinuxMemoryChecker singleton initialization is

```
#ifdef PRESTO_MEMORY_CHECKER_TYPE
#if PRESTO_MEMORY_CHECKER_TYPE == LINUX_MEMORY_CHECKER
// Initialize singleton for the checker.
folly::Singleton<facebook::presto::PeriodicMemoryChecker> checker(
    []() -> facebook::presto::PeriodicMemoryChecker* {
      return createLinuxMemoryChecker().release();
    });
#endif
#endif // PRESTO_MEMORY_CHECKER_TYPE
```

Reply:
^^ No. You would add the selection of the type to the CMakeLists.txt - it determiens when the LinuxMemoryChecker code is added to the build. Compare to the PRESTO_STATS_REPORTER_TYPE usage in the CMakeLists.txt

```
if(PRESTO_STATS_REPORTER_TYPE)
  add_compile_definitions(PRESTO_STATS_REPORTER_TYPE)
  if(PRESTO_STATS_REPORTER_TYPE STREQUAL "PROMETHEUS")
    add_subdirectory(runtime-metrics)
    target_link_libraries(presto_server presto_prometheus_reporter)
  else()
    message(
      FATAL_ERROR
        "${PRESTO_STATS_REPORTER_TYPE} is not a valid stats reporter name")
  endif()
endif()
```
It wouldn't be macros in the source code.

Reply:
Inside CMakeLists.txt, you need to add `add_compile_definitions(PRESTO_STATS_REPORTER_TYPE=${PRESTO_STATS_REPORTER_TYPE} )`
The cpp file would then be
```
if (PRESTO_MEMORY_CHECKER_TYPE == "LINUX_MEMORY_CHECKER") {
}
```

Reply:
You won't be able to add the if statement I suggested since this is a singleton.
See alternative https://github.com/prestodb/presto/pull/22997/files#r1779013312

Reply:
Ok I was thinking of adding this to CMakeLists.txt, how does this look? I removed the add_compile_definition as per Deepak's suggestion since we wouldn't be using it in the C++ code:
```
if(PRESTO_MEMORY_CHECKER_TYPE)
  # Check if the current platform is Linux and the memory checker type is LINUX_MEMORY_CHECKER.
  if(UNIX AND NOT APPLE AND PRESTO_MEMORY_CHECKER_TYPE STREQUAL "LINUX_MEMORY_CHECKER")
    add_library(presto_linux_memory_checker OBJECT LinuxMemoryChecker.cpp)
    target_link_libraries(presto_server PRIVATE presto_linux_memory_checker)
  else()
    message(FATAL_ERROR "${PRESTO_MEMORY_CHECKER_TYPE} is not a valid memory checker name")
  endif()
endif()
```

Reply:
Will also have to create a DummyMemoryChecker class in PeriodicMemoryChecker.h just like how there's DummyStatsReporter class in StatsReporter.h so DummyMemoryChecker can be used for the singleton when PRESTO_MEMORY_CHECKER_TYPE is not defined.